aip/general/0211.md

```diff
@@ -0,0 +1,55 @@
+---
+id: 211
+state: reviewing
+created: 2021-02-24
+placement:
+  category: design-patterns
+  order: 115
+---
+
+# Authorization checks
+
+The majority of operations, whether reads or writes, require authorization:
+permission to do the thing the user is asking to do. Additionally, it is
+important to be careful how much information is provided to _unauthorized_
+users, since leaking information can be a security concern.
+
+## Guidance
+
+Services **must** check authorization before validating any request, to ensure
+both a secure API surface and a consistent user experience. An operation
+**may** require multiple permissions or preconditions in order to grant
+authorization.
+
+If a request can not pass the authorization check for any reason, the service
+**must** error with `PERMISSION_DENIED`, and the corresponding error message
+**should** look like: "Permission `{p}` denied on resource `{r}` (or it might
+not exist)." This avoids leaking resource existance.
+
+**Note:** This guidance conflicts with the guidance in the HTTP specification,
+which argues for the use of 404 in cases where the user lacks permission to
+know whether or not a resource exists. This variance is because of real-world
+challenges: the practice of "getting not found until you have enough permission
+to get permission denied" is counter-intuitive and increases the difficulty of
+troubleshooting. Additionally, the guidance here is more consistent with most
+other authorization systems.
+
+If it is not possible to determine authorization for a resource because the
+resource does not exist, the service **should** check authorization on the
```

Comment:
I think we need to be more specific about what authorization aspect we should check on the parent resource. I *think* this might boil down to the ability to read children of the parent resource, but I could be wrong.

For example... If I want to do anything involving `/shelves/1/books/2` (delete it, read it, update it, ...), but this book doesn't exist, the determining factor between my result being a 403 vs 404 is whether I'm able to call `ListBooks()` on the parent (`shelves/1`) such that I do have a mechanism by which I can determine if the book in question exists.

A caveat here is that if I have access to create resources belonging to the parent **and** I'm able to choose my own identifiers, I am authorized to determine whether the child exists, and therefore should get the 403 (and not the 404).

In code form:

```python
# Whether or not users choose their own IDs when creating resources
USER_CHOSEN_IDS = True

if not resource.exists():
  if resource.parent.check_authz("list", resource.type):
    return 403
  elif USER_CHOSEN_IDS and resource.parent.check_authz("create", resource.type):
    return 403
  else:
    return 404
```

Reply:
> I think this might boil down to the ability to read children of the parent resource, but I could be wrong.

That sounds right to me.

> A caveat here is that if I have access to create resources belonging to the parent and I'm able to choose my own identifiers, I am authorized to determine whether the child exists

Our internal document says this too, but then also says not to try to do 403-404 logic based on it, basically because it becomes too subtle and is likely to be implemented incorrectly.

Reply:
Reworded.